### PR TITLE
Rhmap 9275 repeating sections page rules

### PR DESCRIPTION
**Pull request by jhellar**
# Motivation
https://issues.jboss.org/browse/RHMAP-9275

### CODE REVIEW

File: lib/impl/refactorRules.js [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+        } else {
+          return true;
+        }
+      });
+    }
+
+    var updatedFieldRules = _.map(form.fieldRules, function(fieldRule) {
+      var filteredConditionalStatements = filterConditionalStatements(fieldRule);
+      var filteredTargets = _.filter(fieldRule.targetField, function(targetField) {
+        if (invalidFields[targetField]) {
```
Comment (fh-qube) [inline lib/impl/refactorRules.js]: MINOR Replace this if-then-else statement by a single return statement. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1126)

File: lib/impl/refactorRules.js [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+  var conn = connections.mongooseConnection;
+  var formModel = models.get(conn, models.MODELNAMES.FORM);
+  var fieldModel = models.get(conn, models.MODELNAMES.FIELD);
+  var fieldRuleModel = models.get(conn, models.MODELNAMES.FIELD_RULE);
+  var pageRuleModel = models.get(conn, models.MODELNAMES.PAGE_RULE);
+
+  formModel.findById(form._id.toString()).populate("pages", "-__v")
+    .populate("pageRules", "-__v")
+    .populate("fieldRules", "-__v")
+    .populate("dataSources.formDataSources", "_id name")//Populating Data Sources
```
Comment (fh-qube) [inline lib/impl/refactorRules.js]: INFO Move this trailing comment on the previous empty line. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3ATrailingComment)

File: lib/impl/refactorRules.js [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
+    .populate("dataTargets", "_id name")//Populating Data Targets
```
Comment (fh-qube) [inline lib/impl/refactorRules.js]: INFO Move this trailing comment on the previous empty line. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3ATrailingComment)

File: lib/impl/refactorRules.js [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+          async.series([
+            updateFormRules,
+            updateOrDeleteRules
+          ], cb);
+        });
+      });
+    });
+
+
+  function scanForInvalidFields(type, rule, callback) {
```
Comment (fh-qube) [inline lib/impl/refactorRules.js]: MINOR Function has a complexity of 22 which is greater than 20 authorized. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AFunctionComplexity)

File: lib/impl/refactorRules.js [hunk 5]
```diff
@@ -0,0 +1,384 @@
+var async = require('async');
+var getSectionMap = require('../utils/get-section-map');
+var _ = require('underscore');
+var models = require('../common/models.js')();
+
+/*
+  Checking rule targets are still valid -- if not, rule is deleted
+*/
+module.exports = function refactorRules(connections, form, ruleDeletionFlags, cb) {
```
Comment (fh-qube) [inline lib/impl/refactorRules.js]: MINOR Function has a complexity of 105 which is greater than 20 authorized. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AFunctionComplexity)
MINOR This function has 376 lines, which is greater than the 150 lines authorized. Split it into smaller functions. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS138)

File: lib/impl/refactorRules.js [hunk 6]
```diff
@@ -0,0 +1,384 @@
...
+          });
+          if (targetInDifferrentSection) {
+            rulesFlaggedForDeletion[fieldRule._id.toString()] = true;
+          }
+        }
+      });
+    }
+  }
+
+  function updateOrDeleteRules(cb) {
```
Comment (fh-qube) [inline lib/impl/refactorRules.js]: MINOR Function has a complexity of 29 which is greater than 20 authorized. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AFunctionComplexity)

File: lib/impl/refactorRules.js [hunk 7]
```diff
@@ -0,0 +1,384 @@
...
+          fieldExistsInForm = true;
+        }
+        var invalidFieldList = _.filter(page.fields, function(field) {
+          var currentFieldId = field._id.toString();
+          if (currentFieldId === fieldOrPageIdToCheck) {
+            //Current field exists
+            fieldExistsInForm = true;
+
+            //Field is admin only, therefore it is an invalid target for a rule.
+            if (field.adminOnly) {
```
Comment (fh-qube) [inline lib/impl/refactorRules.js]: MINOR Replace this if-then-else statement by a single return statement. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1126)

File: lib/impl/refactorRules.js [hunk 8]
```diff
@@ -0,0 +1,384 @@
...
+              return cb();
+            }
+          });
+        }, cb);
+      }
+    ], cb);
+  }
+
+  //Need to delete any rules that are to be deleted
+  function updateFormRules(cb) {
```
Comment (fh-qube) [inline lib/impl/refactorRules.js]: MINOR Function has a complexity of 23 which is greater than 20 authorized. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AFunctionComplexity)

**Comment by fh-qube** [comment]
SonarQube analysis reported 9 issues:
* MINOR 7 minor
* INFO 2 info

Watch the comments in this conversation to review them.

File: lib/impl/refactorRules.js [hunk 9]
```diff
@@ -0,0 +1,387 @@
...
+          return targetSectionId !== sectionIds[0];
+        });
+        if (targetInDifferrentSection) {
+          rulesFlaggedForDeletion[fieldRule._id.toString()] = true;
+        }
+      }
+    });
+  }
+
+  function updateOrDeleteRules(cb) {
```
Comment (fh-qube) [inline lib/impl/refactorRules.js]: MINOR Function has a complexity of 29 which is greater than 20 authorized. [rule](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AFunctionComplexity)

**Comment by fh-qube** [comment]
SonarQube analysis reported 9 issues:
* MINOR 7 minor
* INFO 2 info

Watch the comments in this conversation to review them.

**Comment by jhellar** [comment]
@witmicko should be refactored now. Also moved whole `refactorRules` function to separate file -> SonarQube warnings.

**Pull request merged**